Gesture interface merge

mars979 opened the pull request.

rpavlik commented on inc/osvr/Common/RegisteredStringMap.h:
```diff
@@ -0,0 +1,159 @@
+namespace osvr {
+namespace common {
+
+#define MAX_REGISTRY_SIZE 100
```
Nooo! Big code guideline violation. (And use an enlargeable structure instead of having a fixed size)

mars979 replied:
Fixed

rpavlik commented on inc/osvr/Common/RegisteredStringMap.h:
```diff
@@ -0,0 +1,159 @@
+
+#define MAX_REGISTRY_SIZE 100
+
+    typedef char EntryName[100];
```
Nooo! Big code guideline violation - use std::string!

mars979 replied:
Fixed

rpavlik commented on inc/osvr/Common/RegisteredStringMap.h:
```diff
@@ -0,0 +1,159 @@
+
+    /// idea from the
+    /// http://www.ilikebigbits.com/blog/2014/5/6/type-safe-identifiers-in-c
+    template <class Tag, class impl> class ID {
```
I'd actually move this out into its own file as a typesafe handle, and use a policy/traits class for invalid values, etc.

rpavlik commented on inc/osvr/Common/RegisteredStringMap.h:
```diff
@@ -0,0 +1,159 @@
+        friend bool operator==(ID a, ID b) { return a.m_val == b.m_val; }
+        friend bool operator!=(ID a, ID b) { return a.m_val != b.m_val; }
+        */
+        impl m_val;
```
should be private.

rpavlik commented on inc/osvr/Common/RegisteredStringMap.h:
```diff
@@ -0,0 +1,159 @@
+        // explicit operator impl() const { return m_val; }
+
+        // Implicit conversion to get back the impl
+        operator impl() const { return m_val; }
```
This shouldn't be here - should have a get() method or something otherwise we lose type safety.

rpavlik commented on inc/osvr/Common/RegisteredStringMap.h:
```diff
@@ -0,0 +1,159 @@
+        StringID id;
+    } StringRegistry;
+
+    typedef std::vector<StringRegistry> RegistryList;
```
Why not a map? This is a lot of data structures, are there tests?

mars979 replied:
Fixed to use map. Haven't made the tests, that's on todo list

rpavlik commented on inc/osvr/Common/SerializationTraits.h:
```diff
@@ -510,6 +511,15 @@ namespace common {
+        struct SimpleStructSerialization<StringID>
+            : SimpleStructSerializationBase {
+            template <typename F, typename T> static void apply(F &f, T &val) {
+                f(val.m_val);
```
kind of looks like you're accessing a private member: I'd suggest having a member function named `value()` or similar that returns a reference.

rpavlik commented on inc/osvr/PluginKit/GestureInterfaceC.h:
```diff
@@ -0,0 +1,82 @@
+*/
+OSVR_PLUGINKIT_EXPORT
+OSVR_ReturnCode
+osvrDeviceGestureReportData(OSVR_IN_PTR OSVR_GestureDeviceInterface iface,
```
The point in the gesture string map is that the user's code requests a gesture by name once, and holds on to that ID  using it instead of strings. If we use strings each time, we'd do nearly as well to just skip the string mapping. So, need another API method here to get a gesture ID. (API-wise, looks a bit like getting a message type ID)

mars979 replied:
The name to ID map is kept inside the gesture component right now, which looks it up before sending to client. The gesture names get mapped to ID only once but you still need to look up ID by name every time you send it.

rpavlik replied:
OK, so that's going to be between somewhat inefficient and very inefficient.

mars979 replied:
Okay, so we would probably need another data structure in gesture component, like a hash table to provide fast access to IDs, maybe stl's unordered_map?

rpavlik replied:
Unless you have a really good reason not to (type safety sadly not counting given that this is a C API), I'd say that the plugins and clients have access to the numeric IDs directly - that's how the rest of the system was designed and what I had foreseen.

rpavlik commented on inc/osvr/Util/ClientCallbackTypesC.h:
```diff
@@ -200,6 +200,22 @@ typedef struct OSVR_EyeTrackerBlinkReport {
+
+/** @brief Report type for a gesture event */
+typedef struct OSVR_GestureReport {
+    OSVR_GestureName gestureName;
```
Can we just return the local gesture ID here? Then we don't have issues with lifetime of strings, etc.

mars979 replied:
So right now the client doesn't have access to local gesture map to look up name. It is kept inside the gesture remote handler. Did you want to have the client apps be able to access the local map directly?

rpavlik commented on inc/osvr/Util/ClientReportTypesC.h:
```diff
@@ -204,6 +204,45 @@ typedef struct OSVR_EyeTrackerBlinkReport {
+#define OSVR_GESTURE_CLOSED_HAND "ClosedHand"
+
+/** @brief Report type for a gesture event */
+typedef struct OSVR_GestureReport {
```
This is duplicated in `ClientCallbackTypesC.h`

mars979 replied:
Fixed, removed dups from `ClientCallbackTypesC.h`

rpavlik commented on src/osvr/Client/GestureRemoteFactory.cpp:
```diff
@@ -0,0 +1,161 @@
+namespace osvr {
+namespace client {
+
+    class NetworkGestureRemoteHandler : public RemoteHandler {
```
the word "network" here is not useful/needed

mars979 replied:
Fixed

rpavlik commented on src/osvr/Client/GestureRemoteFactory.cpp:
```diff
@@ -0,0 +1,161 @@
+                    util::time::TimeValue const &timestamp) {
+                    m_handleGesture(data, timestamp);
+                });
+            gesture->registerGestureMapHandler(
```
I'd prefer all the map handling to take place in the Component, to keep the logic centralized.

mars979 replied:
Not sure what you meant here. So the gesture component has a separate message that contains updated map of gesture names to IDs which client receives and "synchronizes" with its own.

rpavlik commented on inc/osvr/PluginKit/GestureInterfaceC.h:
```diff
@@ -0,0 +1,82 @@
+OSVR_ReturnCode
+osvrDeviceGestureConfigure(OSVR_INOUT_PTR OSVR_DeviceInitOptions opts,
+                           OSVR_OUT_PTR OSVR_GestureDeviceInterface *iface,
+                           OSVR_IN OSVR_ChannelCount numSensors
```
Do we actually support multiple sensors for gesture per device?

mars979 replied:
That was my understanding that you can have gestures for both hands at least like in LeapMotion https://developer.leapmotion.com/documentation/cpp/devguide/Leap_Hand.html?proglang=cpp
And MYO armband allows multiple devices to be connected and interacted (so 1 armband on each hand is plausible)
https://developer.thalmic.com/docs/api_reference/platform/multiple-myos_8cpp-example.html
Looking at the spec now we defined a gesture not limited to just hands, so you can also have head or other part of the body perform a gesture, then using Kinect you can (in theory) have more than 2 sensors in a device.

rpavlik replied:
OK, that's what I thought. Then we need to think about how best to expose the particular gestures in the path tree and let applications subscribe to just one - whether we just do that as a filter on the client end, or if we modify the path tree to support "sub-sensors" - something like `/plugin/device/gesture/0/pinch`

rpavlik replied:
For now I'd say just a client library filter, have a way to register a callback optionally for just one gesture. The catch is that this means state is weird, since it will tell you the last gesture event to come through.

rpavlik commented on src/osvr/Client/GestureRemoteFactory.cpp:
```diff
@@ -0,0 +1,161 @@
+
+    class NetworkGestureRemoteHandler : public RemoteHandler {
+      public:
+        NetworkGestureRemoteHandler(vrpn_ConnectionPtr const &conn,
```
Is there a way to listen just for a single gesture's callbacks?

rpavlik commented on src/osvr/Client/GestureRemoteFactory.cpp:
```diff
@@ -0,0 +1,161 @@
+            }
+            report.sensor = data.sensor;
+            report.state = data.gestureState;
+            report.gestureName = gestureName.c_str();
```
Can't do this - c_str() returns a temporary, so you can't expect it to hang around.

rpavlik commented on src/osvr/Client/PureClientContext.cpp:
```diff
@@ -117,6 +120,7 @@ namespace client {
         ButtonRemoteFactory(m_vrpnConns).registerWith(m_factory);
         ImagingRemoteFactory(m_vrpnConns).registerWith(m_factory);
         EyeTrackerRemoteFactory(m_vrpnConns).registerWith(m_factory);
+        GestureRemoteFactory(m_vrpnConns).registerWith(m_factory);
```
This should be the only change committed in this file.

rpavlik commented on inc/osvr/Util/ClientReportTypesC.h:
```diff
@@ -204,6 +204,45 @@ typedef struct OSVR_EyeTrackerBlinkReport {
+#define OSVR_GESTURE_COMPLETE (0)
+
+/** @brief a list of pre-set gestures available (can be expanded) */
+#define OSVR_GESTURE_SWIP_LEFT "SwipeLeft"
```
Is there a reason that these are spelled `SWIP` (no `E`)?

mars979 replied:
Fixed the typos

rpavlik commented on src/osvr/Common/GestureComponent.cpp:
```diff
@@ -0,0 +1,236 @@
+        m_gestureNameMap.registerStringID(OSVR_GESTURE_CLOSED_HAND);
+    }
+
+    void GestureComponent::recordParent(Parent &dev) {
```
This is the purpose of having `m_parentSet()` - so you don't override this method if you want to hook in your behavior.

mars979 replied:
Fixed

rpavlik commented on src/osvr/Common/GestureComponent.cpp:
```diff
@@ -0,0 +1,236 @@
+    GestureComponent::m_handleGestureMapRecord(void *userdata,
+                                               vrpn_HANDLERPARAM p) {
+        auto self = static_cast<GestureComponent *>(userdata);
+        auto bufwrap = ExternalBufferReadingWrapper<unsigned char>(
```
See the changes I made recently to clean this up from the eye tracking merge.

mars979 replied:
Fixed. I must have missed it because I changed the first message type but not the second one.

zachkinstner commented:
Notes about Leap Motion gestures: https://github.com/OSVR/OSVR-Leap-Motion/issues/5

rpavlik commented:
The places where an Interface pointer is being held, should instead be using an `InternalInterfaceOwner`, which was just added to master recently.

mars979 commented:
This pull request is ready to be merged

rpavlik commented:
OK - blocked on #218 - after that is finished (unit tests) and merged, this can be rebased on top of the new master. Most of all the "little" pieces that are independent have been pulled out and merged separately, so there will be ample use of `git mergetool` during that rebase, but it should result in a more digestable diff. I'd also suggest squashing related commits down into a single commit, especially early in the history where a lot of things changed since then.

mars979 commented:
I complete unit tests merged #218 yesterday and rebased gesture branch on top of master. Also I fixed a few things that wouldn't let it build and some compiler errors.

rpavlik commented:
Will require fix to adjust for https://github.com/OSVR/OSVR-Core/commit/0a63fd5e091b4b615607f6ddde78acb6f2d49f2d once that branch is merged. Sorry...

rpavlik commented:
If this adds any additional typesafeids beyond those in stringids already in master, those will need to be adjusted for https://github.com/OSVR/OSVR-Core/commit/477c662c5ee093d4d139e5ffe18eabfa27f3fafb
